Workspace/src/science_controller/src/science.py

```diff
@@ -0,0 +1,210 @@
+#!/usr/bin/env python
+from sensor_msgs.msg import Joy
+from std_msgs.msg import UInt8
+from can_msgs.msg import Frame
+import rospy
+import struct
+
+
+class Science:
+    """
+A to open flap, B to close flap. One joy stick to control the spin and
+the other joy stick to control the vertical movement. X to read temp sensor
+and Y to read EC sensor. RT to move rack and pinion system up for
+temp probe motor. RB to move rack and pinion system down for temp
+probe motor.
+Author: Zameer Bharwani
+"""
+
+    def __init__(self):
+
+        self.init_Auger = rospy.init_node("ScienceController", anonymous=True)
+        self.sub_switches = rospy.Subscriber("/science_limit_switches", UInt8,
+                                             self.callback_switches)
+        self.sub_xbox = rospy.Subscriber("/joy", Joy, self.callback_xbox)
+        self.pub = rospy.Publisher("/science", Frame, queue_size=20)
+        self.upper_switch = 0
+        self.lower_switch = 0
+        self.temperature_switch = 0
+        self.speed_dir = 0
+        self.speed_dir_spin = 0
+        self.position = 999
+        self.which_sensor = 999
+        self.temp_sensor = 0
+        self.ID_limit = 420
+        self.ID_spin = 421
+        self.ID_temp_sensor = 422
+        self.ID_flap = 423
+        self.ID_read_sensor = 510
+
+    def callback_switches(self, switches):
+
+        self.upper_switch = (switches.data) & 0x1
+        # check to see if the upper switch is being pressed
+
+        self.lower_switch = (switches.data >> 1) & 0x1
+        # check to see if the lower switch is being pressed
+
+        self.temperature_switch = (switches.data >> 2) & 0x1
+        # check to see if limit switch for temp probe is being pressed
+
+    def callback_xbox(self, joy):
+
+        self.speed_dir_spin = joy.axes[4]
+
+        if self.upper_switch == 1:  # upper limit switch pressed
+
+            if joy.axes[1] <= 0:  # allowing you to move down, but not up
+                self.speed_dir = joy.axes[1]
+
+            else:
+                self.speed_dir = 0  # if still trying to move up do nothing
+
+        elif self.lower_switch == 1:  # lower limit switch pressed
+
+            if joy.axes[1] >= 0:  # allowing you to move up, but not down
+                self.speed_dir = joy.axes[1]
+
+            else:
+                self.speed_dir = 0  # if still trying to move down do nothing
+
+        else:
+            self.speed_dir = joy.axes[1]
+            # if neither sensor is pressed,
+            # movement in either direction is allowed
+
+        shifter = rospy.get_param('budge')  # motor increment value
+
+        if self.temperature_switch == 1:
```

Comment:
switch to constant velocity, just send -1 or 1 depending on button pressed, if nothing pressed then send 0